massa-api/src/proxy.rs

```diff
@@ -0,0 +1,212 @@
+    }
+}
+
+impl<S> Layer<S> for MassaProxyGetRequestLayer {
+    type Service = MassaProxyGetRequest<S>;
+
+    fn layer(&self, inner: S) -> Self::Service {
+        MassaProxyGetRequest::new(inner)
+            .expect("Path already validated in ProxyGetRequestLayer; qed")
+    }
+}
+
+/// Proxy `GET /path` requests to the specified RPC method calls.
+///
+/// # Request
+///
+/// The `GET /path` requests are modified into valid `POST` requests for
+/// calling the RPC method. This middleware adds appropriate headers to the request
+///
+/// # Response
+///
+/// The response of the RPC method is stripped down to contain only the method's
+/// response, removing any RPC 2.0 spec logic regarding the response' body.
+#[derive(Debug, Clone)]
+pub struct MassaProxyGetRequest<S> {
+    inner: S,
+}
+
+impl<S> MassaProxyGetRequest<S> {
+    /// Creates a new [`MassaProxyGetRequest`].
+    pub fn new(inner: S) -> Result<Self, RpcError> {
+        Ok(Self { inner })
+    }
+}
+
+impl<S> Service<Request<Body>> for MassaProxyGetRequest<S>
+where
+    S: Service<Request<Body>, Response = Response<Body>>,
+    S::Response: 'static,
+    S::Error: Into<Box<dyn Error + Send + Sync>> + 'static,
+    S::Future: Send + 'static,
+{
+    type Response = S::Response;
+    type Error = Box<dyn Error + Send + Sync + 'static>;
+    type Future =
+        Pin<Box<dyn Future<Output = Result<Self::Response, Self::Error>> + Send + 'static>>;
+
+    #[inline]
+    fn poll_ready(&mut self, cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
+        self.inner.poll_ready(cx).map_err(Into::into)
+    }
+
+    fn call(&mut self, mut req: Request<Body>) -> Self::Future {
+        let uri = req.uri();
+        let path = uri.path();
+
+        let mut modify = false;
+        let mut params = ArrayParams::new();
+
+        if req.method() == Method::GET
+            && (path.eq(URI_ADDRESSES)
+                || path.eq(URI_BLOCKS)
+                || path.eq(URI_ENDORSEMENTS)
+                || path.eq(URI_OPERATIONS))
+        {
+            if let Some(query) = req.uri().query() {
+                // If this line is reached
+                // Extract params list from query params
+
+                let decoded_query = decode(query)
+                    .expect(&*format!("Error when decode query for GET {}", path))
+                    .to_string();
+
+                for param in decoded_query.split('&') {
+                    // example param for addresses:"ids=["Address1", "Address2"]"
+                    let kv: Vec<&str> = param.splitn(2, '=').collect();
+                    if kv.len() == 2 {
+                        let value: Value = serde_json::from_str(kv[1]).unwrap_or_else(|_| {
+                            panic!("Error when deserialize params query for GET {}", path)
+                        });
+                        params
+                            .insert(value)
+                            .expect("Error when insert into ArrayParams");
```

Comment:
If we could avoid `expect` and handle the error

Reply:
for errors i can directly return an http response to the caller each time a fonction can fail or create a struct like ```MassaProxyError``` or ```MassaProxyGetError``` who do the job for better error management. 
It depends on the future evolution of the proxy, I don’t know if it’s excessive or no
what do you think @aoudiamoncef ?